// yadda/examples/component-browser/components/acuminous-yadda/lib/Macro.js
"use strict";

var fn = require('./fn');
var $ = require('./Array');
var Context = require('./Context');
var RegularExpression = require('./RegularExpression');
var EventBus = require('./EventBus');

// Understands how to invoke a step
var Macro = function(signature, parsed_signature, macro, macro_context) {

    /* jslint shadow: true */
    var signature = normalise(signature);
    var signature_pattern = new RegularExpression(parsed_signature.pattern);
    var macro = macro || fn.async_noop;
    var event_bus = EventBus.instance();

    this.is_identified_by = function(other_signature) {
        return signature == normalise(other_signature);
    };

    this.can_interpret = function(step) {
        return signature_pattern.test(step);
    };

    this.interpret = function(step, scenario_context, next) {
        var context = new Context().merge(macro_context).merge(scenario_context);
        convert(signature_pattern.groups(step), function(err, args) {
            if (err) return next(err);
            event_bus.send(EventBus.ON_EXECUTE, { step: step, ctx: context.properties, pattern: signature_pattern.toString(), args: args });
            fn.invoke(macro, context.properties, args.concat(next));
        });
    };

    this.levenshtein_signature = function() {
        return signature_pattern.without_expressions();
    };

    this.toString = function() {
        return signature;
    };

    function normalise(signature) {
        return new RegExp(signature).toString();
    }

    function convert(args, next) {
        var index = 0;
        return $(parsed_signature.converters).collect(function(converter) {
            return converter.bind.apply(converter, [null].concat(args.slice(index, index += converter.length - 1)));
        }).collect_async(function(converter, index, callback) {
            return converter(callback);
        }, next);
    }
};

module.exports = Macro;
